src/server.c

```diff
@@ -4623,61 +4632,39 @@ int redisSupervisedUpstart(void) {
     return 1;
 }
 
-int redisSupervisedSystemd(void) {
+int redisCommunicateSystemd(const char *sd_notify_msg) {
     const char *notify_socket = getenv("NOTIFY_SOCKET");
-    int fd = 1;
-    struct sockaddr_un su;
-    struct iovec iov;
-    struct msghdr hdr;
-    int sendto_flags = 0;
+    int (*dl_sd_notify)(int unset_environment, const char *state);
+    char *error;
+    void *handle;
 
     if (!notify_socket) {
         serverLog(LL_WARNING,
                 "systemd supervision requested, but NOTIFY_SOCKET not found");
         return 0;
     }
 
-    if ((strchr("@/", notify_socket[0])) == NULL || strlen(notify_socket) < 2) {
-        return 0;
-    }
-
-    serverLog(LL_NOTICE, "supervised by systemd, will signal readiness");
-    if ((fd = socket(AF_UNIX, SOCK_DGRAM, 0)) == -1) {
+    handle = dlopen("libsystemd.so", RTLD_LAZY);
```

Comment:
This is not a very good idea for various reasons. It will require systems to install the development packages to get the .so symlink to the current SONAME (say libsystemd.so.0). This also disregards any ABI, as we are using the non-versioned filename, which once libsystemd bumps its SOVERSION to say 1, it can happily crash or misbehave, etc. It also makes it very hard to spot when this needs to be updated, and no way to prevent loading the wrong library.

The only sane and safe way to use an external library in an optional way would be to create some kind of redis plugin which dynamically links at build-time against libsystemd.so, and then redis can safely use dlopen on its own plugin, because its ABI is guaranteed to match when built together, and the libsystemd ABI is respected because we are always linking against a versioned shared library.